src/ORM/FieldType/DBComposite.php

```diff
@@ -270,10 +270,7 @@ public function setField($field, $value, $markChanged = true)
     {
         $this->objCacheClear();
 
-        // Non-db fields get assigned as normal properties
         if (!$this->hasField($field)) {
-            parent::setField($field, $value);
```

Comment:
There's no way to __get() dynamic properties out of DBComposite because DBComposite::hasField() will always return false

Reply:
If we're going to explicitly not allow setting/getting dynamic data on a `DBComposite` we should throw an exception here, or people will keep running into issues where they think they've set data, but they only ever get `null` when they try to _get_ the data back.

Reply:
This change basically has nothing to do with the changes adding get/set/hasDynmaicData() below, it's just the fact that this piece of code within DBComposite is now functionally useless.  If it's confusing I'll just remove it from the PR

Reply:
The alternative is to keep this in and add a `$this->hasDynamicData($field) ||` to `hasField()`, and a `if ($this->hasDynamicData($field)) { return parent::getField($field); }` to `getField()`.

I'd lean more towards the latter (i.e. restoring support for dynamic data in composite field, and therefore reducing potential upgrade pain) but I don't feel particularly strong either way.

If we don't restore support for dynamic properties we need a blurb in the changelog about it, since it's a breaking change in behaviour.

Whichever solution we go with we need to highlight this scenario in the changelog anyway, actually - since we explicitly tell people to overload `ViewableData::hasField()` we need to explain that they'll need to update their implementations if they want to retain dynamic data support.

Reply:
> restoring support for dynamic data in composite field

I'm not entirely sure if it actually worked properly before, hasField() never looked at dynamic data, so .. maybe it worked via the convoluted magic? Not sure... https://github.com/silverstripe/silverstripe-framework/issues/10684 - was relying on `$this->owner->hasProperty('focuspoint_object')` ... so likely it didn't work.

I think throwing an exception saying use `$this->setDynamicField($field, $value);` is probably the way to go here.

Reply:
Updated

Reply:
`hasField` returned false, but you could still get the value because pup saw that there was a property with that name so it skipped the call to `__get()` and just returned the value outright.

Reply:
Throwing a clear exception is a valid solution though, all good so long as there's a related docs pr

Reply:
OK true. That's a, wonderfully un-intuitive API.  Give its half-backed nature I think just removing it and telling people to use an actual API is the way to go here.

Reply:
docs PR (also linked on the main issue) https://github.com/silverstripe/developer-docs/pull/166